core/src/main/java/org/apache/iceberg/BaseTransaction.java

```diff
@@ -554,7 +554,7 @@ public TableMetadata refresh() {
     @Override
     @SuppressWarnings("ConsistentOverrides")
     public void commit(TableMetadata underlyingBase, TableMetadata metadata) {
-      if (underlyingBase != current) {
+      if (!underlyingBase.metadataFileLocation().equals(current.metadataFileLocation())) {
```

Comment:
This is a weaker requirement. Before, a retry would be triggered if anything has changed and replaced the actual object. I don't think that this is technically incorrect, but it isn't gaining anything. The original check is more strict about triggering a retry and I think it is more safe.

In addition, this should not be included with the REST implementation changes.

Reply:
My concern is that without this change this PR could break existing workflows where `refresh` is being called and is currently a no-op. Since we will now actually change the metadata object regardless of whether or not there is any semantic change.

Reply:
I don't like this check as it currently is because it's very specific to past catalog implementations rather than matching the actual semantic intent. The stronger requirement isn't (or at least shouldn't be) a sign that a retry should take place and it leads us to have to add workarounds to RESTTableOperations.

 ```!Objects.equals(current.metadataFileLocation(), response.metadataLocation()))```
 
 So in that case we only pass this check because we explicitly avoid replacing the object in another part of the code base. That seems a bit messy to me. If the check is good enough for RESTTableOperations we should probably just have it here instead.

Reply:
This is a debate for a different PR. It doesn't have anything to do with fixing REST catalog behavior.

@RussellSpitzer, that is the intent. The safest option for transactions is to consider _any_ change a change that triggers rebuilding the transaction changes.

For other code paths that don't change the current `TableMetadata` object, that is because we prefer to keep the existing copy, which may have snapshots and other classes that cache table state. It isn't that we're careful about not replacing the object in those places so that we can use a faster check here. These are both defensive decisions.